packages/core/src/lib/light_push/light_push.ts

```diff
@@ -16,61 +16,123 @@ import { pipe } from "it-pipe";
 import { Uint8ArrayList } from "uint8arraylist";
 
 import { StreamManager } from "../stream_manager/index.js";
+import { selectOpenConnection } from "../stream_manager/utils.js";
 
-import { PushRpc } from "./push_rpc.js";
+import { PushRpcV2 } from "./push_rpc.js";
+import { PushRpcV3 } from "./push_rpc_v3.js";
+import { isSuccess as isV3Success } from "./utils.js";
 import { isRLNResponseError } from "./utils.js";
 
 const log = new Logger("light-push");
 
-export const LightPushCodec = "/vac/waku/lightpush/2.0.0-beta1";
+export const LightPushCodecV2 = "/vac/waku/lightpush/2.0.0-beta1";
+export const LightPushCodecV3 = "/vac/waku/lightpush/3.0.0";
+export const LightPushCodecs = [LightPushCodecV3, LightPushCodecV2];
 export { PushResponse };
 
-type PreparePushMessageResult = ThisOrThat<"query", PushRpc>;
+type PreparePushMessageResult = ThisOrThat<
+  "query",
+  PushRpcV2 | PushRpcV3,
+  "error",
+  LightPushError
+>;
```

Comment:
I think this is prone to error because you are not leveraging the type system to ensure you are operating with the right protocol.